partners_api/taxi_engine.cpp

```diff
@@ -0,0 +1,203 @@
+
+  if (m_requestId != requestId)
+    return;
+
+  m_errors.emplace_back(type, code);
+
+  DecrementRequestCount();
+}
+
+void ResultMaker::MakeResult(uint64_t requestId) const
+{
+  SuccessfullCallback successCallback;
+  ErrorCallback errorCallback;
+  ProvidersContainer providers;
+  ErrorsContainer errors;
+
+  {
+    std::lock_guard<std::mutex> lock(m_mutex);
+
+    if (m_requestId != requestId || m_requestsCount != 0)
+      return;
+
+    successCallback = m_successCallback;
+    errorCallback = m_errorCallback;
+    providers = m_providers;
+    errors = m_errors;
+  }
+
+  if (providers.empty())
+    return errorCallback(errors, requestId);
+
+  return successCallback(providers, requestId);
+}
+
+void ResultMaker::DecrementRequestCount()
+{
+  CHECK_GREATER(m_requestsCount, 0, ());
+  --m_requestsCount;
+}
+
+Engine::Engine()
+{
+  m_enabledCountries = {{Provider::Type::Yandex, {"Russian Federation"}}};
+  m_disabledCountries = {{Provider::Type::Uber, {"Russian Federation"}}};
+
+  m_apis.emplace_back(Provider::Type::Yandex, my::make_unique<yandex::Api>());
+  m_apis.emplace_back(Provider::Type::Uber, my::make_unique<uber::Api>());
+}
+
+/// Requests list of available products. Returns request identificator immediately.
+uint64_t Engine::GetAvailableProducts(ms::LatLon const & from, ms::LatLon const & to,
+                                      storage::TCountriesVec const & countryIds,
+                                      SuccessfullCallback const & successFn,
+                                      ErrorCallback const & failFn)
+{
+  auto const reqId = ++m_requestId;
+  auto const maker = m_maker;
+
+  maker->Reset(reqId, m_apis.size(), successFn, failFn);
+  for (auto const & api : m_apis)
+  {
+    auto type = api.m_type;
+
+    if (IsCountryDisabled(type, countryIds) || !IsCountryEnabled(type, countryIds))
+    {
+      maker->DecrementRequestCount(reqId);
+      maker->MakeResult(reqId);
+      continue;
+    }
+
+    auto const productCallback = [type, maker, reqId](std::vector<Product> const & products)
+    {
+      maker->ProcessProducts(reqId, type, products);
+      maker->MakeResult(reqId);
+    };
+
+    auto const errorCallback = [type, maker, reqId](ErrorCode const code)
+    {
+      maker->ProcessError(reqId, type, code);
+      maker->MakeResult(reqId);
+    };
+
+    api.m_api->GetAvailableProducts(from, to, productCallback, errorCallback);
+  }
+
+  return reqId;
+}
+
+/// Returns link which allows you to launch some taxi app.
+RideRequestLinks Engine::GetRideRequestLinks(Provider::Type type, std::string const & productId,
+                                             ms::LatLon const & from, ms::LatLon const & to) const
+{
+  auto const it = FindByProviderType(type, m_apis.cbegin(), m_apis.cend());
+
+  CHECK(it != m_apis.cend(), ());
+
+  return it->m_api->GetRideRequestLinks(productId, from, to);
+}
+
+bool Engine::IsCountryDisabled(Provider::Type type, storage::TCountriesVec const & countryIds) const
```

Comment:
Why IsCountryDisabled? Why not AllCountriesDisabled? The latter name is more relevant with the code below, IMHO.